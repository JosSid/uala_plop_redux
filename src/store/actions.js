import {
  ADS_LOADED,
  AUTH_LOGIN_FAILURE,
  AUTH_LOGIN_REQUEST,
  AUTH_LOGIN_SUCCES,
  AUTH_LOGOUT,
  TAGS_LOADED,
  UI_RESET_ERROR,
} from './types';

export const authLoginSucces = () => ({
  type: AUTH_LOGIN_SUCCES,
});

export const authLoginRequest = () => ({
  type: AUTH_LOGIN_REQUEST,
});

export const authLoginFailure = (error) => ({
  type: AUTH_LOGIN_FAILURE,
  payload: error,
  error: true,
});

export const authLogin = (credentials) => {
  return async function (dispatch, getState, {api}) {
    try {
      dispatch(authLoginRequest());
      const accessToken = await api.auth.login(credentials);

      dispatch(authLoginSucces());
      return accessToken;
    } catch (err) {
      dispatch(authLoginFailure(err));
      throw err;
    }
  };
};

export const authLogout = () => ({
  type: AUTH_LOGOUT,
});

export const adsLoaded = (ads) => ({
  type: ADS_LOADED,
  payload: ads,
});

export const tagsLoaded = (tags) => ({
  type: TAGS_LOADED,
  payload: tags,
});

export const uiResetError = () => ({
  type: UI_RESET_ERROR,
});
