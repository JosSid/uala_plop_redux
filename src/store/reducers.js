//  { ESQUEMA DE ESTADO
//      auth: true/false
//      ads: []
//      tags: []
//  }

import { ADS_LOADED, AUTH_LOGIN, AUTH_LOGOUT, TAGS_LOADED } from "./types";

const defaultState = {
    auth: false,
    ads: [],
    tags: []
};

export function auth(state = defaultState.auth, action) {
    if(action.type === AUTH_LOGIN){
        return true;
    };
    if(action.type === AUTH_LOGOUT){
        return false;
    };
    
    return state;
};

export function ads(state = defaultState.ads, action) {
    if(action.type === ADS_LOADED){
        return action.payload;
    };

    return state;
};

export function tags(state = defaultState.tags, action) {
    if(action.type === TAGS_LOADED){
        return action.payload
    };

    return state
};

// export default function reducer( state = defaultState, action) {
//     return {
//         auth: auth(state.auth, action),
//         ads: ads(state.ads, action)}
// };