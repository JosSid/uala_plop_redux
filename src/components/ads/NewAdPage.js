import { useState, useEffect } from 'react';
import {FormField} from '../common/formField/FormField.js';
import Button from '../common/Button.js';
import ErrorDisplay from '../common/error/errorDisplay/ErrorDisplay.js';
import Spinner from '../common/spinner/Spinner.js';
import styles from './NewAdPage.module.css';
import { useDispatch, useSelector } from 'react-redux';
import { getListTags, getUi } from '../../store/selectors.js';
import { tagsLoad, createAd, uiResetError } from '../../store/actions.js';

const NewAdPage = () => {
  const [name, setName] = useState('');
  const [sale, setSale] = useState(true);
  const [price, setPrice] = useState('');
  const [tags, setTags] = useState([]);
  const [photo, setPhoto] = useState(null);
  const {error, isFetching} = useSelector(getUi)
  const listTags = useSelector(getListTags)
  const dispatch = useDispatch()
    
  const handleChangeName = (event) => setName(event.target.value);
  const handleChangeSale = (event) => setSale(JSON.parse(event.target.value));
  const handleChangePrice = (event) => setPrice(parseInt(event.target.value));
  const handleChangeTags = (event) => {
    const selectedTags = event.target.selectedOptions;
    const finallyTags = Array.from(selectedTags).map((e) => e.value);
    setTags(finallyTags);
  };

  const handlePhoto = (event) => {
    setPhoto(event.target.files[0]);
  };

  const isButtonEnabled = () => name && price && tags.length > 0;

  const resetError = () => dispatch(uiResetError());

  const handleSubmit = async (event) => {
    event.preventDefault();
    const formData = new FormData();
    formData.append('name', name);
    formData.append('sale', sale);
    formData.append('price', price);
    photo && formData.append('photo', photo);
    formData.append('tags', tags);
  
    dispatch(createAd(formData))

  };

  useEffect(() => {
    const getNewTags = () => {
      dispatch(tagsLoad())
      };
    getNewTags();
  }, [dispatch]);

  return (
    <div className={styles.newAdPage__container}>
      {isFetching && <Spinner />}
      <h1>Create Advertisment</h1>
      <form data-testid="login-form" className={styles.form__container} onSubmit={handleSubmit}>
        <FormField
          type='text'
          name='name'
          label='Title Advertisment'
          onChange={handleChangeName}
          value={name}
          placeholder='Required field*'
        />
        <fieldset>
          <legend>For sale or Wanted :</legend>
          <label htmlFor='sale'>For Sale</label>
          <input
            type='radio'
            name='sale'
            id='sale'
            value={true}
            onChange={handleChangeSale}
            checked={sale}
          />
          <label htmlFor='wanted'>Wanted</label>
          <input
            type='radio'
            name='sale'
            id='wanted'
            onChange={handleChangeSale}
            value={false}
            checked={!sale}
          />
        </fieldset>
        <FormField
          type='number'
          name='price'
          label='Price'
          onChange={handleChangePrice}
          value={price}
          placeholder='Required field*'
          min='0'
        />
        <label htmlFor='tags'>Tags</label>
        <select
          name='tags'
          id='tags'
          value={tags}
          onChange={handleChangeTags}
          multiple={true}
          size={listTags.length + 1}
        >
          <optgroup label='Required field*'>
            {listTags.map((e) => (
              <option key={e} name={e} value={e}>
                {e}
              </option>
            ))}
          </optgroup>
        </select>
        <label htmlFor='photo'>Upload picture</label>
        <p className={styles.recomendation__size}>
          *Recomended size 300px / 300px
        </p>
        <input
          onChange={handlePhoto}
          type='file'
          name='photo'
          id='photo'
          photo={photo}
        />
        {isButtonEnabled() && (
          <Button
            type='submit'
            className='loginForm-submit'
            disabled={!isButtonEnabled()}
          >
            Create Advertisment
          </Button>
        )}
      </form>
      {error && <ErrorDisplay error={error} resetError={resetError} />}
    </div>
  );
};

export default NewAdPage;
