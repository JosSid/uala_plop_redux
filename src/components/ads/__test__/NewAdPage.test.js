import NewAdPage from "../NewAdPage";
import { fireEvent, render, screen} from '@testing-library/react';
import { Provider } from 'react-redux';
import { defaultState } from '../../../store/reducers.js';
import '@testing-library/jest-dom/extend-expect';
import { createAd } from '../../../store/actions.js';
jest.mock('../../../store/actions.js')

describe('NewAdPage', () => {
    const store = {
        getState: () => defaultState,
        dispatch: () => {},
        subscribe: () => {},
      };

      const renderComponent = () => 
      render(
        <Provider store={store}>
          <NewAdPage />
        </Provider>
      );
    test('SnapShot', () => {
        const { container } = renderComponent();
        expect(container).toMatchSnapshot();
    });
    test('Should dispatch "createAd" action', async () => {
        renderComponent();
        const title = 'Car';
        const price = 100;
        const tags = []

        const nameInput = screen.getByLabelText(/^Title/);
        const priceInput = screen.getByLabelText(/Price/);
        const tagsInput = screen.getByLabelText(/Tags/);

        fireEvent.change(nameInput, { target: { value: title } });
        fireEvent.change(priceInput,  { target: { value: price } });
        fireEvent.change(tagsInput, { target: { value: tags } });
        const submitButton = screen.getByText('Create Advertisment');

        expect(submitButton).toBeEnabled();


       fireEvent.click(submitButton);

       const formElement = await screen.findByTestId('login-form');
        expect(formElement).toHaveFormValues({
            name: 'Car',
            sale: 'true',
            price: 100,
            tags: tags
        });
        fireEvent.submit(formElement);
        expect(createAd).toHaveBeenCalled();
    });
});