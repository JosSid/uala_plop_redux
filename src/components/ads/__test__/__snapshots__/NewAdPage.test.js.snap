// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAdPage SnapShot 1`] = `
<div>
  <div
    class="newAdPage__container"
  >
    <h1>
      Create Advertisment
    </h1>
    <form
      class="form__container"
      data-testid="login-form"
    >
      <div
        class="formField"
      >
        <label
          class="formField__label"
        >
          <span>
            Title Advertisment
          </span>
          <input
            class="formField__input"
            name="name"
            placeholder="Required field*"
            type="text"
            value=""
          />
        </label>
      </div>
      <fieldset>
        <legend>
          For sale or Wanted :
        </legend>
        <label
          for="sale"
        >
          For Sale
        </label>
        <input
          checked=""
          id="sale"
          name="sale"
          type="radio"
          value="true"
        />
        <label
          for="wanted"
        >
          Wanted
        </label>
        <input
          id="wanted"
          name="sale"
          type="radio"
          value="false"
        />
      </fieldset>
      <div
        class="formField"
      >
        <label
          class="formField__label"
        >
          <span>
            Price
          </span>
          <input
            class="formField__input"
            min="0"
            name="price"
            placeholder="Required field*"
            type="number"
            value=""
          />
        </label>
      </div>
      <label
        for="tags"
      >
        Tags
      </label>
      <select
        id="tags"
        multiple=""
        name="tags"
        size="1"
      >
        <optgroup
          label="Required field*"
        />
      </select>
      <label
        for="photo"
      >
        Upload picture
      </label>
      <p
        class="recomendation__size"
      >
        *Recomended size 300px / 300px
      </p>
      <input
        id="photo"
        name="photo"
        type="file"
      />
    </form>
  </div>
</div>
`;
